### PR TITLE
Changes to support Create table

### PR DESCRIPTION
**Pull request by scgkiran**

### CODE REVIEW

File: src/from_substrait.cpp [hunk 1]
```diff
@@ -621,6 +622,95 @@ shared_ptr<Relation> SubstraitToDuckDB::TransformSetOp(const substrait::Rel &sop
...
+	default:
+		throw NotImplementedException("Unsupported DDL operation");
+	}
+}
+
+shared_ptr<Relation> SubstraitToDuckDB::TransformDdlOpCreate(const substrait::Rel &sop) {
+	auto &sddl = sop.ddl();
+	auto ddl_op = sddl.op();
+
+	if (ddl_op != substrait::DdlRel::DdlOp::DdlRel_DdlOp_DDL_OP_CREATE &&
```
Comment (jacques-n) [inline src/from_substrait.cpp:643]: unless I'm misreading, this looks like dead code.

File: src/from_substrait.cpp [hunk 2]
```diff
@@ -621,6 +622,95 @@ shared_ptr<Relation> SubstraitToDuckDB::TransformSetOp(const substrait::Rel &sop
...
+		auto table_name = nobj.names(table_idx);
+		string schema_name;
+		if (table_idx > 0) {
+			schema_name = nobj.names(0);
+		}
+
+		auto &col_names = sddl.table_schema().names();
+		auto col_types = sddl.table_schema().struct_().types();
+		if (col_names.size() != col_types.size()) {
+			throw NotImplementedException("Column names and types count do not match");
```
Comment (jacques-n) [inline src/from_substrait.cpp:666]: seems like wrong error type.

File: src/from_substrait.cpp [hunk 3]
```diff
@@ -621,6 +622,95 @@ shared_ptr<Relation> SubstraitToDuckDB::TransformSetOp(const substrait::Rel &sop
...
+		table_desc->schema = schema_name;
+
+		std::cout << "Creating table1 " << schema_name << " " << table_name << std::endl;
+		shared_ptr<TableRelation> table = make_shared_ptr<TableRelation>(con.context, std::move(table_desc));
+		auto create_table_rel = table->CreateRel(schema_name, table_name, false);
+		std::cout << "Creating table2" << std::endl;
+		return create_table_rel;
+	}
+	default:
+		throw NotImplementedException("Only CreateTable is supported ");
```
Comment (jacques-n) [inline src/from_substrait.cpp:686]: report what this is. E.g. XYZ not supported. Saying what is supported when we hit an exception for something else is hard to debug unless we know what else we hit.

File: src/from_substrait.cpp [hunk 4]
```diff
@@ -621,6 +622,95 @@ shared_ptr<Relation> SubstraitToDuckDB::TransformSetOp(const substrait::Rel &sop
...
+		}
+		auto table_idx = nobj.names_size() - 1;
+		auto table_name = nobj.names(table_idx);
+		string schema_name;
+		if (table_idx > 0) {
+			schema_name = nobj.names(0);
+		}
+
+		auto input = TransformOp(swrite.input());
+		return input->CreateRel(schema_name, table_name, false);
```
Comment (jacques-n) [inline src/from_substrait.cpp:707]: i think the false here should have a comment and TODO:

I'm guessing it is something like "create or replace" vs "create".

File: src/from_substrait.cpp [hunk 5]
```diff
@@ -700,6 +794,18 @@ shared_ptr<Relation> SubstraitToDuckDB::TransformRootOp(const substrait::RelRoot
 		}
 	}
 
+	if (sop.input().rel_type_case() == substrait::Rel::RelTypeCase::kWrite) {
+		auto write = sop.input().write();
+		switch (write.op()) {
+		case substrait::WriteRel::WriteOp::WriteRel_WriteOp_WRITE_OP_CTAS: {
+			const auto create_table = static_cast<CreateTableRelation *>(child.get());
+			auto proj = make_shared_ptr<ProjectionRelation>(create_table->child, std::move(expressions), aliases);
+			return proj->CreateRel(create_table->schema_name, create_table->table_name, false);
```
Comment (jacques-n) [inline src/from_substrait.cpp:803]: todo on false?

File: src/include/to_substrait.hpp [hunk 6]
```diff
@@ -39,6 +39,8 @@ class DuckDBToSubstrait {
 	static vector<string> DepthFirstNames(const LogicalType &type);
 	static void DepthFirstNamesRecurse(vector<string> &names, const LogicalType &type);
 
+        static substrait::Expression_Literal ToExpressionLiteral(const substrait::Expression &expr);
```
Comment (jacques-n) [inline src/include/to_substrait.hpp:42]: weird formatting?

File: src/include/to_substrait.hpp [hunk 7]
```diff
@@ -52,12 +54,15 @@ class DuckDBToSubstrait {
...
+	substrait::Rel *TransformExpressionGet(LogicalOperator &dop);
 	substrait::Rel *TransformCrossProduct(LogicalOperator &dop);
 	substrait::Rel *TransformUnion(LogicalOperator &dop);
 	substrait::Rel *TransformDistinct(LogicalOperator &dop);
 	substrait::Rel *TransformExcept(LogicalOperator &dop);
 	substrait::Rel *TransformIntersect(LogicalOperator &dop);
-	static substrait::Rel *TransformDummyScan();
+	substrait::Rel *TransformCreateTable(LogicalOperator &dop);
+        substrait::Rel *TransformDropTable(LogicalOperator &dop);
+        static substrait::Rel *TransformDummyScan();
```
Comment (jacques-n) [inline src/include/to_substrait.hpp:65]: weird spacing.

File: src/to_substrait.cpp [hunk 8]
```diff
@@ -1,23 +1,25 @@
...
+#include "duckdb/parser/constraints/not_null_constraint.hpp"
 #include "duckdb/planner/expression/list.hpp"
 #include "duckdb/planner/filter/conjunction_filter.hpp"
 #include "duckdb/planner/filter/constant_filter.hpp"
 #include "duckdb/planner/joinside.hpp"
 #include "duckdb/planner/operator/list.hpp"
+#include "duckdb/planner/operator/logical_set_operation.hpp"
 #include "duckdb/planner/table_filter.hpp"
 #include "duckdb/storage/statistics/base_statistics.hpp"
-#include "duckdb/catalog/catalog_entry/duck_table_entry.hpp"
```
Comment (jacques-n) [inline src/to_substrait.cpp:14]: try to avoid arbitrary reordering. let's figure out what duckdb uses and stick with that.

File: src/to_substrait.cpp [hunk 9]
```diff
@@ -1313,6 +1315,43 @@ substrait::Rel *DuckDBToSubstrait::TransformGet(LogicalOperator &dop) {
...
 
+substrait::Expression_Literal DuckDBToSubstrait::ToExpressionLiteral(const substrait::Expression &expr) {
+	substrait::Expression_Literal literal_field;
+	switch (expr.rex_type_case())
+	{
+	case substrait::Expression::kLiteral:
+		literal_field = expr.literal();
+		break;
+	case substrait::Expression::kCast:
+		literal_field = ToExpressionLiteral(expr.cast().input());
```
Comment (jacques-n) [inline src/to_substrait.cpp:1326]: this looks wrong. why can we/should we ignore the cast. if it is right, let's add  a comment. (but I'm not clear how it could be write).

File: src/to_substrait.cpp [hunk 10]
```diff
@@ -1391,6 +1430,90 @@ substrait::Rel *DuckDBToSubstrait::TransformIntersect(LogicalOperator &dop) {
...
 }
 
+substrait::Rel *DuckDBToSubstrait::TransformCreateTable(LogicalOperator &dop) {
+	auto rel = new substrait::Rel();
+	auto &create_table = dop.Cast<LogicalCreateTable>();
+	auto &create_info = create_table.info.get()->Base();
+
+	auto schema = new substrait::NamedStruct();
+	auto type_info = new substrait::Type_Struct();
+	type_info->set_nullability(substrait::Type_Nullability_NULLABILITY_REQUIRED);
```
Comment (jacques-n) [inline src/to_substrait.cpp:1440]: why is this required?

File: src/to_substrait.cpp [hunk 11]
```diff
@@ -1391,6 +1430,90 @@ substrait::Rel *DuckDBToSubstrait::TransformIntersect(LogicalOperator &dop) {
...
+		named_object->add_names(create_info.schema);
+		named_object->add_names(create_info.table);
+		return rel;
+	}
+
+	// This is create table as select
+	substrait::Rel *input = nullptr;
+	switch (create_table.children[0]->type)
+	{
+	case LogicalOperatorType::LOGICAL_PROJECTION:
```
Comment (jacques-n) [inline src/to_substrait.cpp:1467]: why do we need this here? It seems like we would simply take the input and convert using standard methods, no matter the root node type.

File: src/to_substrait.cpp [hunk 12]
```diff
@@ -1391,6 +1430,90 @@ substrait::Rel *DuckDBToSubstrait::TransformIntersect(LogicalOperator &dop) {
...
+		input = TransformProjection(*create_table.children[0]);
+		break;
+	default:
+		throw NotImplementedException("Create table with more than one child");
```
Comment (jacques-n) [inline src/to_substrait.cpp:1471]: can we turn this into a message that would be clearer to an end user? What would generate this situation?

File: src/to_substrait.cpp [hunk 13]
```diff
@@ -1391,6 +1430,90 @@ substrait::Rel *DuckDBToSubstrait::TransformIntersect(LogicalOperator &dop) {
...
+
+	auto ddl = rel->mutable_ddl();
+	switch (drop_info.type) {
+	case CatalogType::TABLE_ENTRY:
+		ddl->set_object(substrait::DdlRel::DDL_OBJECT_TABLE);
+		break;
+	case CatalogType::VIEW_ENTRY:
+		ddl->set_object(substrait::DdlRel::DDL_OBJECT_VIEW);
+	default:
+		throw NotImplementedException("Drop type not implemented");
```
Comment (jacques-n) [inline src/to_substrait.cpp:1502]: report type seen, don't assume that only drop lands here. will send people on wild goose chase/confusion in future.

File: src/to_substrait.cpp [hunk 14]
```diff
@@ -1391,6 +1430,90 @@ substrait::Rel *DuckDBToSubstrait::TransformIntersect(LogicalOperator &dop) {
...
+	}
+	if (drop_info.if_not_found == OnEntryNotFound::THROW_EXCEPTION) {
+		ddl->set_op(substrait::DdlRel::DdlOp::DdlRel_DdlOp_DDL_OP_DROP_IF_EXIST);
```
Comment (jacques-n) [inline src/to_substrait.cpp:1505]: this seems like it is backwards, no?

File: test/c/test_substrait_c_api.cpp [hunk 15]
```diff
@@ -50,3 +54,39 @@ TEST_CASE("Test C Get and To Json-Substrait API", "[substrait-api]") {
...
+  REQUIRE_NO_FAIL(con.Query("SELECT * from t1"));
+
+  auto json_plan = con.GetSubstraitJSON("create table t2 as SELECT * FROM t1");
+  result = con.FromSubstraitJSON(json_plan);
+
+  proto = con.GetSubstrait("select * from t2 limit 2");
+  result = con.FromSubstrait(proto);
+
+  REQUIRE(CHECK_COLUMN(result, 0, {"john", "jane"}));
+  REQUIRE(CHECK_COLUMN(result, 1, {25, 21}));
```
Comment (jacques-n) [inline test/c/test_substrait_c_api.cpp:91]: let's add some additional tests similar to other items (python?)

**Review by jacques-n — COMMENTED** [review]
does this PR depend on the build PR? If so, let's get that merged. If not, let's extract this off of that so we can get this merged.

File: src/from_substrait.cpp (hunk 5 above)
Comment (scgkiran) [reply src/from_substrait.cpp:803]: false is optional flag for temporary. I am removing it.

File: src/to_substrait.cpp (hunk 10 above)
Comment (scgkiran) [reply src/to_substrait.cpp:1440]: removed it.

File: src/to_substrait.cpp (hunk 9 above)
Comment (anshuldata) [reply src/to_substrait.cpp:1326]: I see your point. I think I did a mistake here.
We need to get literal from cast expression here. This is required because virtual table values are literal ([here](https://github.com/substrait-io/substrait/blob/1f3354d9f0f8b4425e98623f34d1e4578e2142bd/proto/substrait/algebra.proto#L91))

My understanding is, we need to convert this cast expression to literal based on duckdb type system. Let me know if this understanding is correct ?

File: src/from_substrait.cpp (hunk 1 above)
Comment (scgkiran) [reply src/from_substrait.cpp:643]: yes,  getting rid of this.

File: src/from_substrait.cpp (hunk 2 above)
Comment (scgkiran) [reply src/from_substrait.cpp:666]: dead code removing

File: src/from_substrait.cpp (hunk 3 above)
Comment (scgkiran) [reply src/from_substrait.cpp:686]: dead code removing. Following the advice at other places.

File: src/from_substrait.cpp (hunk 4 above)
Comment (scgkiran) [reply src/from_substrait.cpp:707]: last argument is optional for  temporary (table?). Removing the arg.

File: src/include/to_substrait.hpp (hunk 6 above)
Comment (scgkiran) [reply src/include/to_substrait.hpp:42]: fixing it, will check after raising new PR

File: src/to_substrait.cpp (hunk 8 above)
Comment (scgkiran) [reply src/to_substrait.cpp:14]: removed it.

File: src/to_substrait.cpp (hunk 11 above)
Comment (scgkiran) [reply src/to_substrait.cpp:1467]: fixed it.

File: src/to_substrait.cpp (hunk 12 above)
Comment (scgkiran) [reply src/to_substrait.cpp:1471]: made this as an internalException. I think it should never hit this path.

File: src/to_substrait.cpp (hunk 13 above)
Comment (scgkiran) [reply src/to_substrait.cpp:1502]: removing the drop code.

File: src/to_substrait.cpp (hunk 14 above)
Comment (scgkiran) [reply src/to_substrait.cpp:1505]: removing the drop code.

File: src/include/to_substrait.hpp (hunk 7 above)
Comment (scgkiran) [reply src/include/to_substrait.hpp:65]: fixing it, will check after raising new PR

File: src/to_substrait.cpp (hunk 9 above)
Comment (anshuldata) [reply src/to_substrait.cpp:1326]: We need to handle it as Project<Expression>(Virtual table). This will be handled as separate PR

File: test/c/test_substrait_c_api.cpp (hunk 15 above)
Comment (scgkiran) [reply test/c/test_substrait_c_api.cpp:91]: Added C tests. Working on python tests.
Raised a new [PR](https://github.com/scgkiran/duckdb-substrait/pull/2) stacked on top of [PR for cmake changes](https://github.com/sundeck-io/duckdb-substrait/pull/4)

**Comment by scgkiran** [comment]
Opened in PR #7
